### PR TITLE
Don't create cookies when disabled

### LINKED ISSUE #17892: Don't create test cookies when cookies are disabled
refs https://forum.matomo.org/t/force-tracking-without-cookies-and-disable-cookies-not-working/42563/10
refs https://github.com/matomo-org/tag-manager/pull/158

When having cookies disabled in the Matomo configuration variable, and even when in privacy anonoymisation settings cookies are force disabled, then Matomo JS tracker will still create a cookie and then delete it again right away.

This is because `setVisitorCookie` is called in https://github.com/matomo-org/matomo/blob/4.4.1/js/piwik.js#L4830 before any other plugins etc can disable it.

And because tag manager does `new Tracker(url, idsite)` it already has the idsite at time of tracker instance creation and therefore it does set a cookie in https://github.com/matomo-org/matomo/blob/f57a5313f7923c60e55c2b101351bccaae287330/js/piwik.js#L3341 which is then deleted again. 

The problem is when clicking eg in Chrome on the SSL certificate and then Cookies then this cookie still appears there and it looks to visitors as if cookies were used.

There used to be this fix for it in Tag Manager: https://github.com/matomo-org/tag-manager/pull/158

We then believed this temporary workaround was no longer needed as not having the idSite in constructor can cause other side effects. It looks like it is still an issue though. In https://github.com/matomo-org/tag-manager/pull/261 we changed it back and thought it was working now but it isn't.

Maybe a workaround to fix this issue could be to call the visitor cookie method around here https://github.com/matomo-org/matomo/blob/f57a5313f7923c60e55c2b101351bccaae287330/js/piwik.js#L6938 after triggering the `TrackerSetup` event.

To reproduce this create a new tracker instance where you create a URL and idSite.

we should comment in https://forum.matomo.org/t/force-tracking-without-cookies-and-disable-cookies-not-working/42563/12 once fixed

### PR DESCRIPTION
**Pull request by justinvelluppillai**
### Description:

Fixes #17892 

Supersedes #18135 (created a few new PRs to get around submodule issues).

This PR moves visitorId cookie creation a bit later, ie only when needed for `trackPageView()` or similar call. VisitorId is now stored internally in the tracker to offset some of the impact of this, but in some cases where multiple trackers are being used there may be some functional differences as implied by the changes I've made to the tests where they only pass when `trackPageView()` is called before other tracker methods.

### Review

* [ ] [Functional review done](https://developer.matomo.org/guides/pull-request-reviews#functional-review-done)
* [ ] [Potential edge cases thought about](https://developer.matomo.org/guides/pull-request-reviews#potential-edge-cases-thought-about) (behavior of the code with strange input, with strange internal state or possible interactions with other Matomo subsystems)
* [ ] [Usability review done](https://developer.matomo.org/guides/pull-request-reviews#usability-review-done) (is anything maybe unclear or think about anything that would cause people to reach out to support)
* [ ] [Security review done](https://developer.matomo.org/guides/security-in-piwik#checklist)
* [ ] [Wording review done](https://developer.matomo.org/guides/pull-request-reviews#translations-wording-review-done)
* [ ] [Code review done](https://developer.matomo.org/guides/pull-request-reviews#code-review-done)
* [ ] [Tests were added if useful/possible](https://developer.matomo.org/guides/pull-request-reviews#tests-were-added-if-usefulpossible)
* [ ] [Reviewed for breaking changes](https://developer.matomo.org/guides/pull-request-reviews#reviewed-for-breaking-changes)
* [ ] [Developer changelog updated if needed](https://developer.matomo.org/guides/pull-request-reviews#developer-changelog-updated-if-needed)
* [ ] [Documentation added if needed](https://developer.matomo.org/guides/pull-request-reviews#documentation-added-if-needed)
* [ ] Existing documentation updated if needed

### CODE REVIEW

**Comment by justinvelluppillai** [comment]
build js

**Comment by justinvelluppillai** [comment]
All relevant tests now passing, so this is ready for review with particular consideration to implications for other plugins or unusual tracking setups.

**Comment by sgiehl** [comment]
I looked through the code changes and imho they are looking fine. The only "issue" I could see is the slightly changed behavior when using multiple trackers. But that actually should only affect using the visitorid in the trackers before a first tracking request was send. To circumvent this, I guess the only possible solution would be to somehow store the initital visitorid in a global scope, that can be accessed in each tracker instance. But not sure if that is needed.
@tsteur do you think we should move that to a global scope?

**Comment by tsteur** [comment]
@sgiehl I don't quite understand the issue. AFAIK, we expect each tracker to have a different visitorId, not the same. But I might remember it wrongly. In what case exactly would it cause an issue? Like how can it be reproduced?

**Comment by justinvelluppillai** [comment]
These tests suggest we want the same visitorId across trackers https://github.com/matomo-org/matomo/blob/e33a147f1c74524ef584ee80fc5c77b8fb6b59aa/tests/javascript/index.php#L2285. There's maybe no issue there but there is a change of behaviour which could be removed by setting a global visitorId on the page.

**Comment by tsteur** [comment]
They are only the same if they point to the same URL and site. 

If there are different trackers used (pointing to a different site, matomo url, or cookie prefix), then they should be different. Which we can also see in  https://github.com/matomo-org/matomo/blob/e33a147f1c74524ef584ee80fc5c77b8fb6b59aa/tests/javascript/index.php#L2303-L2312

Meaning it shouldn't be global if I see this right.


Because in https://github.com/matomo-org/matomo/blob/e33a147f1c74524ef584ee80fc5c77b8fb6b59aa/tests/javascript/index.php#L2291 we are tracking a page view, I would have expected the cookie would be set and we shouldn't see a change in behaviour. 

So the test is definitely expecting the right thing, the question is why isn't the `delayedTracker` reading the same visitorId. Maybe it's not returning the correct tracker instance? My random guess might be calling `tracker.removeAllAsyncTrackersButFirst()` in the beginning might fix the issue in case there's a tracker configured differently from a previous test.

**Comment by justinvelluppillai** [comment]
@tsteur that line 2291 I added. The change in behaviour was that tracking a pageview is now needed where before this would have worked without it

**Comment by tsteur** [comment]
@justinvelluppillai if someone did call `getVisitorId` we could also set the cookie if not set yet? This would fix the issue maybe?

And if someone didn't call `getVisitorId` both trackers would still use the same visitor ID as long as they read from the same cookie I would assume?

**Comment by sgiehl** [comment]
@tsteur what about use cases where `getVisitorId` might be called before the permission to set cookies was granted?

**Comment by tsteur** [comment]
@sgiehl I don't think that be a problem that often (using `getVisitorId` and manually creating the tracker)? Do you maybe have a specific implementation example where this be a problem? Do I see this right if `getVisitorId` is called too early it would return an empty visitor ID?

**Comment by justinvelluppillai** [comment]
If we set the cookie when `getVisitorId` is called then we still have the original problem of potentially having cookies set when disabled. `getVisitorId` wouldn't return an empty visitor ID if called early, it would call `loadVisitorIdCookie` which will generate a UUID and return a new visitor ID if no cookie exists yet.

**Comment by tsteur** [comment]
That be fine in that case @justinvelluppillai 👍  At least then people would have a way to fix this issue by calling `getVisitorId` after configuring cookie consent etc. And we improve the situation for 99.8% of the people

**Comment by justinvelluppillai** [comment]
build js

**Comment by justinvelluppillai** [comment]
build js

**Comment by justinvelluppillai** [comment]
build js

**Comment by justinvelluppillai** [comment]
build js

**Comment by justinvelluppillai** [comment]
Nice one thanks @tsteur. I've implemented that approach here and it passes tests without making those changes 👍 

A couple of unrelated tests fail but this is good for review.

**Comment by sgiehl** [comment]
@justinvelluppillai there is also `getVisitorInfo`, which currently might return varying userids until a cookie was set, guess we should set the cookie in that method as well.
Also I'm not sure if it is expected that the visitorId is currently empty until the cookie could be set. But I guess this actually already was the case before.

**Comment by justinvelluppillai** [comment]
build js

**Comment by justinvelluppillai** [comment]
@sgiehl I've added to now set visitor id cookie in `getVisitorInfo` also. And yes, this doesn't change behaviour from before with the visitorId before the cookie is set, except previously the visitorId cookie was set immediately if possible.

**Comment by justinvelluppillai** [comment]
build js

File: js/piwik.js [hunk 1]
```diff
@@ -7071,6 +7074,15 @@ if (typeof window.Matomo !== 'object') {
             });
 
             Matomo.trigger('TrackerSetup', [this]);
+
+            Matomo.addPlugin('TrackerVisitorIdCookie' + uniqueTrackerId, {
+                // if no tracking request was sent we refresh the visitor id cookie on page unload
+                unload: function () {
+                    if (!hasSentTrackingRequestYet) {
+                        setVisitorIdCookie();
```
Comment (sgiehl) [inline js/piwik.js:7082]: Wouldn't this mean, that if the user leaves the page (without previous tracking), but did not yet gave consent a cookie would be set nevertheless?

Comment (justinvelluppillai) [reply js/piwik.js:7082]: My understanding is that the `setCookie` function takes care of that by respecting `configCookiesDisabled` which is set to `true` if consent is required. I could more explicitly add the check here if needed.

Comment (justinvelluppillai) [reply js/piwik.js:7082]: Also remember previously we called `setVisitorIdCookie()` much earlier, so this doesn't change behaviour other than delaying this call.

Comment (sgiehl) [reply js/piwik.js:7082]: was just wondering as you do a check for `(!configCookiesDisabled && !getCookie(getCookieName('id')))` within `getVisitorInfo`, but not here.

Comment (justinvelluppillai) [reply js/piwik.js:7082]: good point, I'll actually remove the check on `getVisitorInfo()` as it is unnecessary.

**Comment by justinvelluppillai** [comment]
build js

**Review by sgiehl — APPROVED** [review]

**Pull request merged**